<a href="https://colab.research.google.com/github/FestusMikhael/PENELITIAN-TA/blob/main/kode%20eksperimen.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
!pip install -q ultralytics timm

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 41.3/41.3 kB 2.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 26.6 MB/s eta 0:00:00


In [ ]:
import os
import torch
import torch.nn as nn
import timm
import yaml
from ultralytics import YOLO
from ultralytics.nn import tasks
import ultralytics.nn.modules.conv as conv_modules

Creating new Ultralytics Settings v0.0.6 file ✅ 
View Ultralytics Settings with 'yolo settings' or at '/root/.config/Ultralytics/settings.json'
Update Settings with 'yolo settings key=value', i.e. 'yolo settings runs_dir=path/to/dir'. For help see https://docs.ultralytics.com/quickstart/#ultralytics-settings.


In [ ]:
print("\n--- Menyalin dan Mengekstrak Dataset ---")
!cp /content/drive/MyDrive/RDD2022_JP_YOLO/RDD2022_JP_YOLO_v2.zip /content/
!unzip -q /content/RDD2022_JP_YOLO_v2.zip -d /content/rdd_data/
print("Dataset berhasil disiapkan di: /content/rdd_data/")


--- Menyalin dan Mengekstrak Dataset ---
Dataset berhasil disiapkan di: /content/rdd_data/


##**YOLOV8N**

In [ ]:
# TAHAP TRAINING MODEL BASELINE: YOLOv8n STANDAR (ORIGINAL)
from ultralytics import YOLO

print("\n--- Memulai Training YOLOv8n Standar (Baseline) ---")
model_standar = YOLO('yolov8n.pt')

model_standar.train(
    data='/content/drive/MyDrive/RDD2022_JP_YOLO/rdd_data_jp_local.yaml', # Path dataset Anda
    epochs=50,
    imgsz=640,
    batch=16,
    name='yolov8n',
    project='yolo_comparison_test'
)

print("\n Training YOLOv8n Standar Selesai!")


--- Memulai Training YOLOv8n Standar (Baseline) ---
Ultralytics 8.4.48 🚀 Python-3.12.13 torch-2.10.0+cu128 CUDA:0 (Tesla T4, 14913MiB)
engine/trainer: agnostic_nms=False, amp=True, angle=1.0, augment=False, auto_augment=randaugment, batch=16, bgr=0.0, box=7.5, cache=False, cfg=None, classes=None, close_mosaic=10, cls=0.5, cls_pw=0.0, compile=False, conf=None, copy_paste=0.0, copy_paste_mode=flip, cos_lr=False, cutmix=0.0, data=/content/drive/MyDrive/RDD2022_JP_YOLO/rdd_data_jp_local.yaml, degrees=0.0, deterministic=True, device=None, dfl=1.5, dnn=False, dropout=0.0, dynamic=False, embed=None, end2end=None, epochs=50, erasing=0.4, exist_ok=False, fliplr=0.5, flipud=0.0, format=torchscript, fraction=1.0, freeze=None, half=False, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, imgsz=640, int8=False, iou=0.7, keras=False, kobj=1.0, line_width=None, lr0=0.01, lrf=0.01, mask_ratio=4, max_det=300, mixup=0.0, mode=train, model=yolov8n.pt, momentum=0.937, mosaic=1.0, multi_scale=0.0, name=yolov8n, nbs=64, 

### Daftar Parameter Training (Ultralytics v8.4.48)

| Kategori | Parameter | Nilai | Deskripsi |
| :--- | :--- | :--- | :--- |
| **Utama** | `epochs` | 50 | Total siklus pelatihan |
| | `imgsz` | 640 | Resolusi input gambar |
| | `batch` | 16 | Jumlah gambar per batch |
| **Optimizer** | `optimizer` | Auto (SGD) | Algoritma optimasi bobot |
| | `lr0` | 0.01 | Initial Learning Rate |
| | `lrf` | 0.01 | Final Learning Rate |
| | `momentum` | 0.937 | Faktor momentum gradien |
| | `weight_decay` | 0.0005 | Penalti pencegahan overfitting |
| **Loss Gain** | `box` | 7.5 | Bobot penalti Box Loss |
| | `cls` | 0.5 | Bobot penalti Classification Loss |
| | `dfl` | 1.5 | Bobot Distribution Focal Loss |
| **Augmentasi** | `mosaic` | 1.0 | Probabilitas penggabungan 4 gambar |
| | `close_mosaic`| 10 | Mematikan mosaic di 10 epoch akhir |
| | `warmup_epochs`| 3.0 | Fase pemanasan awal |
| **Lainnya** | `iou` | 0.7 | Ambang batas Intersection over Union |


In [ ]:
# MENGHITUNG METRIK LENGKAP (YOLOv8n STANDAR)
print("\n--- Menghitung Metrik Lengkap (YOLOv8n Standar) ---")
model_eval = YOLO('/content/runs/detect/yolo_comparison_test/yolov8n/weights/best.pt')

# Lakukan validasi ulang secara singkat
metrik = model_eval.val(data='/content/drive/MyDrive/RDD2022_JP_YOLO/rdd_data_jp_local.yaml')

# Panggil metrik spesifik
print("\n=== HASIL EVALUASI YOLOv8n STANDAR ===")
print(f"mAP@50    : {metrik.box.map50:.4f}")
print(f"mAP@75    : {metrik.box.map75:.4f}")
print(f"mAP@50-95 : {metrik.box.map:.4f}")


--- Menghitung Metrik Lengkap (YOLOv8n Standar) ---
Ultralytics 8.4.48 🚀 Python-3.12.13 torch-2.10.0+cu128 CUDA:0 (Tesla T4, 14913MiB)
Model summary (fused): 73 layers, 3,006,623 parameters, 0 gradients, 8.1 GFLOPs
val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 2585.0±876.4 MB/s, size: 79.1 KB)
val: Scanning /content/rdd_data/content/drive/MyDrive/RDD2022_JP_YOLO/val/labels.cache... 1863 images, 277 backgrounds, 0 corrupt: 100% ━━━━━━━━━━━━ 1863/1863 520.9Mit/s 0.0s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 117/117 5.8it/s 20.1s
                   all       1863       3345      0.591      0.531      0.559       0.26
    longitudinal crack        571        844      0.533      0.466      0.488      0.218
      transverse crack        446        806      0.568      0.396       0.45      0.173
       alligator crack        940       1233      0.667      0.652      0.692      0.368
               pothole        297      

##**YOLOV8N + MOBILENETV3 SMALL**


In [ ]:
# TAHAP 1: CUSTOM LAYER REGISTRATION (LAYER ADAPTER)
class Select_Explicit_Fix(nn.Module):
    def __init__(self, c1, c2, index=0, **kwargs):
        super().__init__()
        self.index = index

    def forward(self, x):
        if isinstance(x, list):
            return x[self.index]
        return x

# Override modul agar YOLO mengenali selector custom
conv_modules.GhostConv = Select_Explicit_Fix
tasks.GhostConv = Select_Explicit_Fix

# TAHAP 2: DEFINISI BACKBONE (MOBILENET V3 SMALL)
MODEL_NAME = 'mobilenetv3_small_100'
CORRECT_INDICES = [2, 3, 4]

class MobileNetV3Backbone(nn.Module):
    def __init__(self, *args, **kwargs):
        super().__init__()
        self.model = timm.create_model(
            MODEL_NAME, features_only=True, out_indices=CORRECT_INDICES, pretrained=True, in_chans=3
        )
    def forward(self, x):
        return self.model(x)

tasks.MobileNetV3Backbone = MobileNetV3Backbone

# Ambil dimensi channel asli dari MobileNet secara otomatis
tmp_model = timm.create_model(MODEL_NAME, features_only=True, out_indices=CORRECT_INDICES, pretrained=False)
tmp_out = tmp_model(torch.randn(1, 3, 640, 640))
REAL_CHANNELS = [x.shape[1] for x in tmp_out]
print(f"Dimensi Channel Backbone: {REAL_CHANNELS}")

# TAHAP 3: KONFIGURASI ARSITEKTUR YAML
yaml_filename = 'yolov8n_mbv3.yaml'

config_content = {
    'nc': 4,
    'scales': {'n': [0.33, 1.00, 1024]},
    'backbone': [
        [-1, 1, 'MobileNetV3Backbone', []],
        [0, 1, 'GhostConv', [REAL_CHANNELS[0], 0]], # P3
        [0, 1, 'GhostConv', [REAL_CHANNELS[1], 1]], # P4
        [0, 1, 'GhostConv', [REAL_CHANNELS[2], 2]], # P5
        [-1, 1, 'SPPF', [256, 5]]
    ],
    'head': [
        [-1, 1, 'nn.Upsample', [None, 2, 'nearest']],
        [[-1, 2], 1, 'Concat', [1]],
        [-1, 3, 'C2f', [64]],

        [-1, 1, 'nn.Upsample', [None, 2, 'nearest']],
        [[-1, 1], 1, 'Concat', [1]],
        [-1, 3, 'C2f', [64]],

        [-1, 1, 'Conv', [64, 3, 2]],
        [[-1, 7], 1, 'Concat', [1]],
        [-1, 3, 'C2f', [128]],

        [-1, 1, 'Conv', [128, 3, 2]],
        [[-1, 4], 1, 'Concat', [1]],
        [-1, 3, 'C2f', [256]],

        [[10, 13, 16], 1, 'Detect', ['nc']]
    ]
}

with open(yaml_filename, 'w') as f:
    yaml.dump(config_content, f, sort_keys=False)

# TAHAP 4: INISIALISASI & TRAINING
print(f"\n--- Memulai Training YOLOv8n + MobileNetV3 ---")
model = YOLO(yaml_filename)

model.train(
    data='/content/drive/MyDrive/RDD2022_JP_YOLO/rdd_data_jp_local.yaml',
    epochs=50,
    imgsz=640,
    batch=16,
    name='yolov8n_mbv3',
    project='yolo_comparison_test'
)

Dimensi Channel Backbone: [24, 48, 576]

--- Memulai Training YOLOv8n + MobileNetV3 ---


/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:93: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


model.safetensors:   0%|          | 0.00/10.2M [00:00<?, ?B/s]

Ultralytics 8.4.49 🚀 Python-3.12.13 torch-2.10.0+cu128 CUDA:0 (Tesla T4, 14913MiB)
engine/trainer: agnostic_nms=False, amp=True, angle=1.0, augment=False, auto_augment=randaugment, batch=16, bgr=0.0, box=7.5, cache=False, cfg=None, classes=None, close_mosaic=10, cls=0.5, cls_pw=0.0, compile=False, conf=None, copy_paste=0.0, copy_paste_mode=flip, cos_lr=False, cutmix=0.0, data=/content/drive/MyDrive/RDD2022_JP_YOLO/rdd_data_jp_local.yaml, degrees=0.0, deterministic=True, device=None, dfl=1.5, dnn=False, dropout=0.0, dynamic=False, embed=None, end2end=None, epochs=50, erasing=0.4, exist_ok=False, fliplr=0.5, flipud=0.0, format=torchscript, fraction=1.0, freeze=None, half=False, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, imgsz=640, int8=False, iou=0.7, keras=False, kobj=1.0, line_width=None, lr0=0.01, lrf=0.01, mask_ratio=4, max_det=300, mixup=0.0, mode=train, model=yolov8n_mbv3.yaml, momentum=0.937, mosaic=1.0, multi_scale=0.0, name=yolov8n_mbv3, nbs=64, nms=False, opset=None, optimize=False, op

  0                  -1  1    927008  MobileNetV3Backbone                          []                            
  1                   0  1         0  Select_Explicit_Fix                          [3, 24, 0]                    
  2                   0  1         0  Select_Explicit_Fix                          [3, 48, 1]                    
  3                   0  1         0  Select_Explicit_Fix                          [3, 576, 2]                   
  4                  -1  1    461888  ultralytics.nn.modules.block.SPPF            [576, 256, 5]                 
  5                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']          
  6             [-1, 2]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
  7                  -1  1     44416  ultralytics.nn.modules.block.C2f             [304, 64, 1]                  
  8                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None

ultralytics.utils.metrics.DetMetrics object with attributes:

ap_class_index: array([0, 1, 2, 3])
box: ultralytics.utils.metrics.Metric object
confusion_matrix: <ultralytics.utils.metrics.ConfusionMatrix object at 0x7819b0f24a40>
curves: ['Precision-Recall(B)', 'F1-Confidence(B)', 'Precision-Confidence(B)', 'Recall-Confidence(B)']
curves_results: [[array([          0,    0.001001,    0.002002,    0.003003,    0.004004,    0.005005,    0.006006,    0.007007,    0.008008,    0.009009,     0.01001,    0.011011,    0.012012,    0.013013,    0.014014,    0.015015,    0.016016,    0.017017,    0.018018,    0.019019,     0.02002,    0.021021,    0.022022,    0.023023,
          0.024024,    0.025025,    0.026026,    0.027027,    0.028028,    0.029029,     0.03003,    0.031031,    0.032032,    0.033033,    0.034034,    0.035035,    0.036036,    0.037037,    0.038038,    0.039039,     0.04004,    0.041041,    0.042042,    0.043043,    0.044044,    0.045045,    0.046046,    0.047047,
          0

In [ ]:
model_eval = 'YOLO'(path/best.pt)
metrik = model_eval.val(path/image001.png)

print (metrik)


In [ ]:
# MENGHITUNG METRIK LENGKAP (YOLOv8n + MobileNetV3)
print("\n--- Menghitung Metrik Lengkap (YOLOv8n + MobilNetv3) ---")
model_eval = YOLO('/content/runs/detect/yolo_comparison_test/yolov8n_mbv3/weights/best.pt')

# Lakukan validasi ulang secara singkat
print("\n--- Menghitung Metrik Lengkap ---")
metrik = model_eval.val(data='/content/drive/MyDrive/RDD2022_JP_YOLO/rdd_data_jp_local.yaml')

# Panggil metrik spesifik
print(f"mAP@50    : {metrik.box.map50:.4f}")
print(f"mAP@75    : {metrik.box.map75:.4f}")
print(f"mAP@50-95 : {metrik.box.map:.4f}")


--- Menghitung Metrik Lengkap (YOLOv8n + MobilNetv3) ---

--- Menghitung Metrik Lengkap ---
Ultralytics 8.4.49 🚀 Python-3.12.13 torch-2.10.0+cu128 CUDA:0 (Tesla T4, 14913MiB)
YOLOv8n_mbv3 summary: 215 layers, 3,006,367 parameters, 0 gradients, 6.1 GFLOPs
val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 2347.4±840.7 MB/s, size: 79.1 KB)
val: Scanning /content/rdd_data/content/drive/MyDrive/RDD2022_JP_YOLO/val/labels.cache... 1863 images, 277 backgrounds, 0 corrupt: 100% ━━━━━━━━━━━━ 1863/1863 868.2Mit/s 0.0s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 117/117 5.6it/s 20.8s
                   all       1863       3345       0.54      0.445      0.471      0.213
    longitudinal crack        571        844      0.527      0.364      0.401      0.166
      transverse crack        446        806      0.468      0.329      0.361      0.128
       alligator crack        940       1233      0.635      0.586      0.638      0.335

### Konfigurasi Training Modifikasi (MobileNetV3 Standar Tanpa Tuning)

| Kategori | Parameter | Nilai | Deskripsi |
| :--- | :--- | :--- | :--- |
| **Utama** | `model` | yolov8n_mbv3.yaml | Arsitektur modifikasi MobileNetV3 |
| | `imgsz` | 640 | Resolusi input gambar |
| | `epochs` | 50 | Total siklus pelatihan |
| | `batch` | 16 | Jumlah sampel per batch |
| **Optimizer** | `optimizer` | Auto (SGD) | Algoritma optimasi bobot |
| | `lr0` | 0.01 | Initial Learning Rate |
| | `momentum` | 0.937 | Faktor momentum gradien |
| | `weight_decay` | 0.0005 | Penalti regularisasi |
| **Loss Gain** | `box` | 7.5 | Bobot penalti Box Loss |
| | `cls` | 0.5 | Bobot penalti Classification Loss |
| | `dfl` | 1.5 | Bobot Distribution Focal Loss |
| **Augmentasi** | `mosaic` | 1.0 | Aktif (Penggabungan 4 gambar) |
| | `close_mosaic`| 10 | Mosaic dimatikan pada 10 epoch terakhir |
| | `warmup_epochs`| 3.0 | Durasi fase pemanasan awal |
| **Lainnya** | `iou` | 0.7 | NMS Threshold (Penyaringan kotak dobel) |

##**HYPERPARAMETER**

In [ ]:
# TAHAP 1: CUSTOM LAYER REGISTRATION (LAYER ADAPTER)
class Select_Explicit_Fix(nn.Module):
    def __init__(self, c1, c2, index=0, **kwargs):
        super().__init__()
        self.index = index

    def forward(self, x):
        if isinstance(x, list):
            return x[self.index]
        return x

# Override modul agar YOLO mengenali selector custom
conv_modules.GhostConv = Select_Explicit_Fix
tasks.GhostConv = Select_Explicit_Fix

# TAHAP 2: DEFINISI BACKBONE (MOBILENET V3 SMALL)
MODEL_NAME = 'mobilenetv3_small_100'
CORRECT_INDICES = [2, 3, 4]

class MobileNetV3Backbone(nn.Module):
    def __init__(self, *args, **kwargs):
        super().__init__()
        self.model = timm.create_model(
            MODEL_NAME, features_only=True, out_indices=CORRECT_INDICES, pretrained=True, in_chans=3
        )
    def forward(self, x):
        return self.model(x)

tasks.MobileNetV3Backbone = MobileNetV3Backbone

# Ambil dimensi channel asli dari MobileNet secara otomatis
tmp_model = timm.create_model(MODEL_NAME, features_only=True, out_indices=CORRECT_INDICES, pretrained=False)
tmp_out = tmp_model(torch.randn(1, 3, 640, 640))
REAL_CHANNELS = [x.shape[1] for x in tmp_out]
print(f"Dimensi Channel Backbone: {REAL_CHANNELS}")

# TAHAP 3: KONFIGURASI ARSITEKTUR YAML
yaml_filename = 'yolov8n_mbv3.yaml'

config_content = {
    'nc': 4,
    'scales': {'n': [0.33, 1.00, 1024]},
    'backbone': [
        [-1, 1, 'MobileNetV3Backbone', []],
        [0, 1, 'GhostConv', [REAL_CHANNELS[0], 0]], # P3
        [0, 1, 'GhostConv', [REAL_CHANNELS[1], 1]], # P4
        [0, 1, 'GhostConv', [REAL_CHANNELS[2], 2]], # P5
        [-1, 1, 'SPPF', [256, 5]]
    ],
    'head': [
        [-1, 1, 'nn.Upsample', [None, 2, 'nearest']],
        [[-1, 2], 1, 'Concat', [1]],
        [-1, 3, 'C2f', [64]],

        [-1, 1, 'nn.Upsample', [None, 2, 'nearest']],
        [[-1, 1], 1, 'Concat', [1]],
        [-1, 3, 'C2f', [64]],

        [-1, 1, 'Conv', [64, 3, 2]],
        [[-1, 7], 1, 'Concat', [1]],
        [-1, 3, 'C2f', [128]],

        [-1, 1, 'Conv', [128, 3, 2]],
        [[-1, 4], 1, 'Concat', [1]],
        [-1, 3, 'C2f', [256]],

        [[10, 13, 16], 1, 'Detect', ['nc']]
    ]
}

with open(yaml_filename, 'w') as f:
    yaml.dump(config_content, f, sort_keys=False)

Dimensi Channel Backbone: [24, 48, 576]


In [ ]:
# TAHAP 4: MANUAL TUNING - SKENARIO 1
print(f"\n--- Memulai Manual Tuning (Learning Rate: 0.001) ---")
model = YOLO('yolov8n_mbv3.yaml')

model.train(
    data='/content/drive/MyDrive/RDD2022_JP_YOLO/rdd_data_jp_local.yaml',
    epochs=10,
    imgsz=640,
    batch=16,
    lr0=0.001,
    name='yolov8n_mbv3_manual_S1',
    project='yolo_tuning_test'
)


--- Memulai Manual Tuning (Learning Rate: 0.001) ---


/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:93: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


model.safetensors:   0%|          | 0.00/10.2M [00:00<?, ?B/s]

Ultralytics 8.4.46 🚀 Python-3.12.13 torch-2.10.0+cu128 CUDA:0 (Tesla T4, 14913MiB)
engine/trainer: agnostic_nms=False, amp=True, angle=1.0, augment=False, auto_augment=randaugment, batch=16, bgr=0.0, box=7.5, cache=False, cfg=None, classes=None, close_mosaic=10, cls=0.5, cls_pw=0.0, compile=False, conf=None, copy_paste=0.0, copy_paste_mode=flip, cos_lr=False, cutmix=0.0, data=/content/drive/MyDrive/RDD2022_JP_YOLO/rdd_data_jp_local.yaml, degrees=0.0, deterministic=True, device=None, dfl=1.5, dnn=False, dropout=0.0, dynamic=False, embed=None, end2end=None, epochs=10, erasing=0.4, exist_ok=False, fliplr=0.5, flipud=0.0, format=torchscript, fraction=1.0, freeze=None, half=False, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, imgsz=640, int8=False, iou=0.7, keras=False, kobj=1.0, line_width=None, lr0=0.001, lrf=0.01, mask_ratio=4, max_det=300, mixup=0.0, mode=train, model=yolov8n_mbv3.yaml, momentum=0.937, mosaic=1.0, multi_scale=0.0, name=yolov8n_mbv3_manual_S1, nbs=64, nms=False, opset=None, optimiz

  0                  -1  1    927008  MobileNetV3Backbone                          []                            
  1                   0  1         0  Select_Explicit_Fix                          [3, 24, 0]                    
  2                   0  1         0  Select_Explicit_Fix                          [3, 48, 1]                    
  3                   0  1         0  Select_Explicit_Fix                          [3, 576, 2]                   
  4                  -1  1    461888  ultralytics.nn.modules.block.SPPF            [576, 256, 5]                 
  5                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']          
  6             [-1, 2]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
  7                  -1  1     44416  ultralytics.nn.modules.block.C2f             [304, 64, 1]                  
  8                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None

ultralytics.utils.metrics.DetMetrics object with attributes:

ap_class_index: array([0, 1, 2, 3])
box: ultralytics.utils.metrics.Metric object
confusion_matrix: <ultralytics.utils.metrics.ConfusionMatrix object at 0x7e0cf6dbf020>
curves: ['Precision-Recall(B)', 'F1-Confidence(B)', 'Precision-Confidence(B)', 'Recall-Confidence(B)']
curves_results: [[array([          0,    0.001001,    0.002002,    0.003003,    0.004004,    0.005005,    0.006006,    0.007007,    0.008008,    0.009009,     0.01001,    0.011011,    0.012012,    0.013013,    0.014014,    0.015015,    0.016016,    0.017017,    0.018018,    0.019019,     0.02002,    0.021021,    0.022022,    0.023023,
          0.024024,    0.025025,    0.026026,    0.027027,    0.028028,    0.029029,     0.03003,    0.031031,    0.032032,    0.033033,    0.034034,    0.035035,    0.036036,    0.037037,    0.038038,    0.039039,     0.04004,    0.041041,    0.042042,    0.043043,    0.044044,    0.045045,    0.046046,    0.047047,
          0

In [ ]:
# TAHAP 4: MANUAL TUNING - SKENARIO 2
print(f"\n--- Eksperimen Skenario 2: Mencari Kecepatan Konvergensi (LR: 0.005) ---")
model = YOLO('yolov8n_mbv3.yaml')

model.train(
    data='/content/drive/MyDrive/RDD2022_JP_YOLO/rdd_data_jp_local.yaml',
    epochs=10,
    imgsz=640,
    batch=16,
    lr0=0.005,       # Menengah antara 0.01 (default) dan 0.001 (S1)
    warmup_epochs=2, # Tancap gas lebih awal
    name='yolov8n_mbv3_manual_S2',
    project='yolo_tuning_test'
)


--- Eksperimen Skenario 2: Mencari Kecepatan Konvergensi (LR: 0.005) ---


/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:93: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


model.safetensors:   0%|          | 0.00/10.2M [00:00<?, ?B/s]

Ultralytics 8.4.46 🚀 Python-3.12.13 torch-2.10.0+cu128 CUDA:0 (Tesla T4, 14913MiB)
engine/trainer: agnostic_nms=False, amp=True, angle=1.0, augment=False, auto_augment=randaugment, batch=16, bgr=0.0, box=7.5, cache=False, cfg=None, classes=None, close_mosaic=10, cls=0.5, cls_pw=0.0, compile=False, conf=None, copy_paste=0.0, copy_paste_mode=flip, cos_lr=False, cutmix=0.0, data=/content/drive/MyDrive/RDD2022_JP_YOLO/rdd_data_jp_local.yaml, degrees=0.0, deterministic=True, device=None, dfl=1.5, dnn=False, dropout=0.0, dynamic=False, embed=None, end2end=None, epochs=10, erasing=0.4, exist_ok=False, fliplr=0.5, flipud=0.0, format=torchscript, fraction=1.0, freeze=None, half=False, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, imgsz=640, int8=False, iou=0.7, keras=False, kobj=1.0, line_width=None, lr0=0.005, lrf=0.01, mask_ratio=4, max_det=300, mixup=0.0, mode=train, model=yolov8n_mbv3.yaml, momentum=0.937, mosaic=1.0, multi_scale=0.0, name=yolov8n_mbv3_manual_S2, nbs=64, nms=False, opset=None, optimiz

  0                  -1  1    927008  MobileNetV3Backbone                          []                            
  1                   0  1         0  Select_Explicit_Fix                          [3, 24, 0]                    
  2                   0  1         0  Select_Explicit_Fix                          [3, 48, 1]                    
  3                   0  1         0  Select_Explicit_Fix                          [3, 576, 2]                   
  4                  -1  1    461888  ultralytics.nn.modules.block.SPPF            [576, 256, 5]                 
  5                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']          
  6             [-1, 2]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
  7                  -1  1     44416  ultralytics.nn.modules.block.C2f             [304, 64, 1]                  
  8                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None

ultralytics.utils.metrics.DetMetrics object with attributes:

ap_class_index: array([0, 1, 2, 3])
box: ultralytics.utils.metrics.Metric object
confusion_matrix: <ultralytics.utils.metrics.ConfusionMatrix object at 0x7f4c2afabda0>
curves: ['Precision-Recall(B)', 'F1-Confidence(B)', 'Precision-Confidence(B)', 'Recall-Confidence(B)']
curves_results: [[array([          0,    0.001001,    0.002002,    0.003003,    0.004004,    0.005005,    0.006006,    0.007007,    0.008008,    0.009009,     0.01001,    0.011011,    0.012012,    0.013013,    0.014014,    0.015015,    0.016016,    0.017017,    0.018018,    0.019019,     0.02002,    0.021021,    0.022022,    0.023023,
          0.024024,    0.025025,    0.026026,    0.027027,    0.028028,    0.029029,     0.03003,    0.031031,    0.032032,    0.033033,    0.034034,    0.035035,    0.036036,    0.037037,    0.038038,    0.039039,     0.04004,    0.041041,    0.042042,    0.043043,    0.044044,    0.045045,    0.046046,    0.047047,
          0

In [ ]:
# TAHAP 4: MANUAL TUNING - SKENARIO 3
print(f"\n--- Eksperimen Skenario 3: Optimizer AdamW & Mosaic Tuning (LR: 0.005) ---")
model = YOLO('yolov8n_mbv3.yaml')

model.train(
    data='/content/drive/MyDrive/RDD2022_JP_YOLO/rdd_data_jp_local.yaml',
    epochs=10,
    imgsz=640,
    batch=16,
    lr0=0.005,
    optimizer='AdamW', # Ganti optimizer
    mosaic=0.5,        # Sesuaikan augmentasi
    warmup_epochs=2,
    name='yolov8n_mbv3_manual_S3',
    project='yolo_tuning_test'
)


--- Eksperimen Skenario 3: Optimizer AdamW & Mosaic Tuning (LR: 0.005) ---


/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:93: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


model.safetensors:   0%|          | 0.00/10.2M [00:00<?, ?B/s]

Ultralytics 8.4.46 🚀 Python-3.12.13 torch-2.10.0+cu128 CUDA:0 (Tesla T4, 14913MiB)
engine/trainer: agnostic_nms=False, amp=True, angle=1.0, augment=False, auto_augment=randaugment, batch=16, bgr=0.0, box=7.5, cache=False, cfg=None, classes=None, close_mosaic=10, cls=0.5, cls_pw=0.0, compile=False, conf=None, copy_paste=0.0, copy_paste_mode=flip, cos_lr=False, cutmix=0.0, data=/content/drive/MyDrive/RDD2022_JP_YOLO/rdd_data_jp_local.yaml, degrees=0.0, deterministic=True, device=None, dfl=1.5, dnn=False, dropout=0.0, dynamic=False, embed=None, end2end=None, epochs=10, erasing=0.4, exist_ok=False, fliplr=0.5, flipud=0.0, format=torchscript, fraction=1.0, freeze=None, half=False, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, imgsz=640, int8=False, iou=0.7, keras=False, kobj=1.0, line_width=None, lr0=0.005, lrf=0.01, mask_ratio=4, max_det=300, mixup=0.0, mode=train, model=yolov8n_mbv3.yaml, momentum=0.937, mosaic=0.5, multi_scale=0.0, name=yolov8n_mbv3_manual_S3, nbs=64, nms=False, opset=None, optimiz

  0                  -1  1    927008  MobileNetV3Backbone                          []                            
  1                   0  1         0  Select_Explicit_Fix                          [3, 24, 0]                    
  2                   0  1         0  Select_Explicit_Fix                          [3, 48, 1]                    
  3                   0  1         0  Select_Explicit_Fix                          [3, 576, 2]                   
  4                  -1  1    461888  ultralytics.nn.modules.block.SPPF            [576, 256, 5]                 
  5                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']          
  6             [-1, 2]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
  7                  -1  1     44416  ultralytics.nn.modules.block.C2f             [304, 64, 1]                  
  8                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None

ultralytics.utils.metrics.DetMetrics object with attributes:

ap_class_index: array([0, 1, 2, 3])
box: ultralytics.utils.metrics.Metric object
confusion_matrix: <ultralytics.utils.metrics.ConfusionMatrix object at 0x7ccaf18d5580>
curves: ['Precision-Recall(B)', 'F1-Confidence(B)', 'Precision-Confidence(B)', 'Recall-Confidence(B)']
curves_results: [[array([          0,    0.001001,    0.002002,    0.003003,    0.004004,    0.005005,    0.006006,    0.007007,    0.008008,    0.009009,     0.01001,    0.011011,    0.012012,    0.013013,    0.014014,    0.015015,    0.016016,    0.017017,    0.018018,    0.019019,     0.02002,    0.021021,    0.022022,    0.023023,
          0.024024,    0.025025,    0.026026,    0.027027,    0.028028,    0.029029,     0.03003,    0.031031,    0.032032,    0.033033,    0.034034,    0.035035,    0.036036,    0.037037,    0.038038,    0.039039,     0.04004,    0.041041,    0.042042,    0.043043,    0.044044,    0.045045,    0.046046,    0.047047,
          0

In [ ]:
# TAHAP 4: MANUAL TUNING - SKENARIO 4 (Augmentation Boost)
print(f"\n--- Eksperimen Skenario 4: Mixup & Copy-Paste (LR: 0.005) ---")
model = YOLO('yolov8n_mbv3.yaml')

model.train(
    data='/content/drive/MyDrive/RDD2022_JP_YOLO/rdd_data_jp_local.yaml',
    epochs=10,
    imgsz=640,
    batch=16,
    lr0=0.005,
    optimizer='SGD',
    warmup_epochs=2,
    mixup=0.2,         # Menambah kesulitan data
    copy_paste=0.1,    # Menambah variasi posisi objek
    degrees=10.0,      # Variasi rotasi
    name='yolov8n_mbv3_manual_S4',
    project='yolo_tuning_test'
)


--- Eksperimen Skenario 4: Mixup & Copy-Paste (LR: 0.005) ---


/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:93: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


model.safetensors:   0%|          | 0.00/10.2M [00:00<?, ?B/s]

Ultralytics 8.4.47 🚀 Python-3.12.13 torch-2.10.0+cu128 CUDA:0 (Tesla T4, 14913MiB)
engine/trainer: agnostic_nms=False, amp=True, angle=1.0, augment=False, auto_augment=randaugment, batch=16, bgr=0.0, box=7.5, cache=False, cfg=None, classes=None, close_mosaic=10, cls=0.5, cls_pw=0.0, compile=False, conf=None, copy_paste=0.1, copy_paste_mode=flip, cos_lr=False, cutmix=0.0, data=/content/drive/MyDrive/RDD2022_JP_YOLO/rdd_data_jp_local.yaml, degrees=10.0, deterministic=True, device=None, dfl=1.5, dnn=False, dropout=0.0, dynamic=False, embed=None, end2end=None, epochs=10, erasing=0.4, exist_ok=False, fliplr=0.5, flipud=0.0, format=torchscript, fraction=1.0, freeze=None, half=False, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, imgsz=640, int8=False, iou=0.7, keras=False, kobj=1.0, line_width=None, lr0=0.005, lrf=0.01, mask_ratio=4, max_det=300, mixup=0.2, mode=train, model=yolov8n_mbv3.yaml, momentum=0.937, mosaic=1.0, multi_scale=0.0, name=yolov8n_mbv3_manual_S4, nbs=64, nms=False, opset=None, optimi

  0                  -1  1    927008  MobileNetV3Backbone                          []                            
  1                   0  1         0  Select_Explicit_Fix                          [3, 24, 0]                    
  2                   0  1         0  Select_Explicit_Fix                          [3, 48, 1]                    
  3                   0  1         0  Select_Explicit_Fix                          [3, 576, 2]                   
  4                  -1  1    461888  ultralytics.nn.modules.block.SPPF            [576, 256, 5]                 
  5                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']          
  6             [-1, 2]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
  7                  -1  1     44416  ultralytics.nn.modules.block.C2f             [304, 64, 1]                  
  8                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None

ultralytics.utils.metrics.DetMetrics object with attributes:

ap_class_index: array([0, 1, 2, 3])
box: ultralytics.utils.metrics.Metric object
confusion_matrix: <ultralytics.utils.metrics.ConfusionMatrix object at 0x7a414d751280>
curves: ['Precision-Recall(B)', 'F1-Confidence(B)', 'Precision-Confidence(B)', 'Recall-Confidence(B)']
curves_results: [[array([          0,    0.001001,    0.002002,    0.003003,    0.004004,    0.005005,    0.006006,    0.007007,    0.008008,    0.009009,     0.01001,    0.011011,    0.012012,    0.013013,    0.014014,    0.015015,    0.016016,    0.017017,    0.018018,    0.019019,     0.02002,    0.021021,    0.022022,    0.023023,
          0.024024,    0.025025,    0.026026,    0.027027,    0.028028,    0.029029,     0.03003,    0.031031,    0.032032,    0.033033,    0.034034,    0.035035,    0.036036,    0.037037,    0.038038,    0.039039,     0.04004,    0.041041,    0.042042,    0.043043,    0.044044,    0.045045,    0.046046,    0.047047,
          0

In [ ]:
# TAHAP 4: MANUAL TUNING - SKENARIO 6 (High Momentum)
model = YOLO('yolov8n_mbv3.yaml')

model.train(
    data='/content/drive/MyDrive/RDD2022_JP_YOLO/rdd_data_jp_local.yaml',
    epochs=10,
    imgsz=640,
    batch=16,
    lr0=0.005,
    optimizer='SGD',
    momentum=0.98,     # Naikkan momentum
    warmup_epochs=2,
    name='yolov8n_mbv3_manual_S6',
    project='yolo_tuning_test'
)

/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:93: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


model.safetensors:   0%|          | 0.00/10.2M [00:00<?, ?B/s]

Ultralytics 8.4.47 🚀 Python-3.12.13 torch-2.10.0+cu128 CUDA:0 (Tesla T4, 14913MiB)
engine/trainer: agnostic_nms=False, amp=True, angle=1.0, augment=False, auto_augment=randaugment, batch=16, bgr=0.0, box=7.5, cache=False, cfg=None, classes=None, close_mosaic=10, cls=0.5, cls_pw=0.0, compile=False, conf=None, copy_paste=0.0, copy_paste_mode=flip, cos_lr=False, cutmix=0.0, data=/content/drive/MyDrive/RDD2022_JP_YOLO/rdd_data_jp_local.yaml, degrees=0.0, deterministic=True, device=None, dfl=1.5, dnn=False, dropout=0.0, dynamic=False, embed=None, end2end=None, epochs=10, erasing=0.4, exist_ok=False, fliplr=0.5, flipud=0.0, format=torchscript, fraction=1.0, freeze=None, half=False, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, imgsz=640, int8=False, iou=0.7, keras=False, kobj=1.0, line_width=None, lr0=0.005, lrf=0.01, mask_ratio=4, max_det=300, mixup=0.0, mode=train, model=yolov8n_mbv3.yaml, momentum=0.98, mosaic=1.0, multi_scale=0.0, name=yolov8n_mbv3_manual_S6, nbs=64, nms=False, opset=None, optimize

  0                  -1  1    927008  MobileNetV3Backbone                          []                            
  1                   0  1         0  Select_Explicit_Fix                          [3, 24, 0]                    
  2                   0  1         0  Select_Explicit_Fix                          [3, 48, 1]                    
  3                   0  1         0  Select_Explicit_Fix                          [3, 576, 2]                   
  4                  -1  1    461888  ultralytics.nn.modules.block.SPPF            [576, 256, 5]                 
  5                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']          
  6             [-1, 2]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
  7                  -1  1     44416  ultralytics.nn.modules.block.C2f             [304, 64, 1]                  
  8                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None

ultralytics.utils.metrics.DetMetrics object with attributes:

ap_class_index: array([0, 1, 2, 3])
box: ultralytics.utils.metrics.Metric object
confusion_matrix: <ultralytics.utils.metrics.ConfusionMatrix object at 0x7a57cd564710>
curves: ['Precision-Recall(B)', 'F1-Confidence(B)', 'Precision-Confidence(B)', 'Recall-Confidence(B)']
curves_results: [[array([          0,    0.001001,    0.002002,    0.003003,    0.004004,    0.005005,    0.006006,    0.007007,    0.008008,    0.009009,     0.01001,    0.011011,    0.012012,    0.013013,    0.014014,    0.015015,    0.016016,    0.017017,    0.018018,    0.019019,     0.02002,    0.021021,    0.022022,    0.023023,
          0.024024,    0.025025,    0.026026,    0.027027,    0.028028,    0.029029,     0.03003,    0.031031,    0.032032,    0.033033,    0.034034,    0.035035,    0.036036,    0.037037,    0.038038,    0.039039,     0.04004,    0.041041,    0.042042,    0.043043,    0.044044,    0.045045,    0.046046,    0.047047,
          0

In [ ]:
# TAHAP 4: MANUAL TUNING - SKENARIO 5 (IMG Size)
print(f"\n--- Memulai Training YOLOv8n + MobileNetV3 (Resolution 800) ---")
model = YOLO('yolov8n_mbv3.yaml')

model.train(
    data='/content/drive/MyDrive/RDD2022_JP_YOLO/rdd_data_jp_local.yaml',
    epochs=10,
    imgsz=800,
    batch=16,
    name='yolov8n_mbv3_manual_S5',
    project='yolo_tuning_test'
)


--- Memulai Training YOLOv8n + MobileNetV3 (Resolution 800) ---


/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:93: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


model.safetensors:   0%|          | 0.00/10.2M [00:00<?, ?B/s]

Ultralytics 8.4.47 🚀 Python-3.12.13 torch-2.10.0+cu128 CUDA:0 (Tesla T4, 14913MiB)
engine/trainer: agnostic_nms=False, amp=True, angle=1.0, augment=False, auto_augment=randaugment, batch=16, bgr=0.0, box=7.5, cache=False, cfg=None, classes=None, close_mosaic=10, cls=0.5, cls_pw=0.0, compile=False, conf=None, copy_paste=0.0, copy_paste_mode=flip, cos_lr=False, cutmix=0.0, data=/content/drive/MyDrive/RDD2022_JP_YOLO/rdd_data_jp_local.yaml, degrees=0.0, deterministic=True, device=None, dfl=1.5, dnn=False, dropout=0.0, dynamic=False, embed=None, end2end=None, epochs=10, erasing=0.4, exist_ok=False, fliplr=0.5, flipud=0.0, format=torchscript, fraction=1.0, freeze=None, half=False, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, imgsz=800, int8=False, iou=0.7, keras=False, kobj=1.0, line_width=None, lr0=0.01, lrf=0.01, mask_ratio=4, max_det=300, mixup=0.0, mode=train, model=yolov8n_mbv3.yaml, momentum=0.937, mosaic=1.0, multi_scale=0.0, name=yolov8n_mbv3_manual_S5, nbs=64, nms=False, opset=None, optimize

  0                  -1  1    927008  MobileNetV3Backbone                          []                            
  1                   0  1         0  Select_Explicit_Fix                          [3, 24, 0]                    
  2                   0  1         0  Select_Explicit_Fix                          [3, 48, 1]                    
  3                   0  1         0  Select_Explicit_Fix                          [3, 576, 2]                   
  4                  -1  1    461888  ultralytics.nn.modules.block.SPPF            [576, 256, 5]                 
  5                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']          
  6             [-1, 2]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
  7                  -1  1     44416  ultralytics.nn.modules.block.C2f             [304, 64, 1]                  
  8                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None

ultralytics.utils.metrics.DetMetrics object with attributes:

ap_class_index: array([0, 1, 2, 3])
box: ultralytics.utils.metrics.Metric object
confusion_matrix: <ultralytics.utils.metrics.ConfusionMatrix object at 0x79ae7d5e6f30>
curves: ['Precision-Recall(B)', 'F1-Confidence(B)', 'Precision-Confidence(B)', 'Recall-Confidence(B)']
curves_results: [[array([          0,    0.001001,    0.002002,    0.003003,    0.004004,    0.005005,    0.006006,    0.007007,    0.008008,    0.009009,     0.01001,    0.011011,    0.012012,    0.013013,    0.014014,    0.015015,    0.016016,    0.017017,    0.018018,    0.019019,     0.02002,    0.021021,    0.022022,    0.023023,
          0.024024,    0.025025,    0.026026,    0.027027,    0.028028,    0.029029,     0.03003,    0.031031,    0.032032,    0.033033,    0.034034,    0.035035,    0.036036,    0.037037,    0.038038,    0.039039,     0.04004,    0.041041,    0.042042,    0.043043,    0.044044,    0.045045,    0.046046,    0.047047,
          0

##**TUNING - 1 (BEST)**


In [ ]:
print(f"\n--- Memulai Training ---")
model = YOLO('yolov8n_mbv3.yaml')

model.train(
    data='/content/drive/MyDrive/RDD2022_JP_YOLO/rdd_data_jp_local.yaml',
    epochs=50,
    imgsz=800,
    batch=16,
    lr0=0.01,
    optimizer='SGD',
    warmup_epochs=3,
    close_mosaic=15,
    label_smoothing=0.1,
    name='yolov8n_mbv3_tuning',
    project='yolo_tuning_test'
)


--- Memulai Training ---


/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:93: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


model.safetensors:   0%|          | 0.00/10.2M [00:00<?, ?B/s]

WARNING ⚠️ 'label_smoothing' is deprecated and will be removed in the future.
Ultralytics 8.4.48 🚀 Python-3.12.13 torch-2.10.0+cu128 CUDA:0 (Tesla T4, 14913MiB)
engine/trainer: agnostic_nms=False, amp=True, angle=1.0, augment=False, auto_augment=randaugment, batch=16, bgr=0.0, box=7.5, cache=False, cfg=None, classes=None, close_mosaic=15, cls=0.5, cls_pw=0.0, compile=False, conf=None, copy_paste=0.0, copy_paste_mode=flip, cos_lr=False, cutmix=0.0, data=/content/drive/MyDrive/RDD2022_JP_YOLO/rdd_data_jp_local.yaml, degrees=0.0, deterministic=True, device=None, dfl=1.5, dnn=False, dropout=0.0, dynamic=False, embed=None, end2end=None, epochs=50, erasing=0.4, exist_ok=False, fliplr=0.5, flipud=0.0, format=torchscript, fraction=1.0, freeze=None, half=False, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, imgsz=800, int8=False, iou=0.7, keras=False, kobj=1.0, line_width=None, lr0=0.01, lrf=0.01, mask_ratio=4, max_det=300, mixup=0.0, mode=train, model=yolov8n_mbv3.yaml, momentum=0.937, mosaic=1.0, multi_s

  0                  -1  1    927008  MobileNetV3Backbone                          []                            
  1                   0  1         0  Select_Explicit_Fix                          [3, 24, 0]                    
  2                   0  1         0  Select_Explicit_Fix                          [3, 48, 1]                    
  3                   0  1         0  Select_Explicit_Fix                          [3, 576, 2]                   
  4                  -1  1    461888  ultralytics.nn.modules.block.SPPF            [576, 256, 5]                 
  5                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']          
  6             [-1, 2]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
  7                  -1  1     44416  ultralytics.nn.modules.block.C2f             [304, 64, 1]                  
  8                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None

ultralytics.utils.metrics.DetMetrics object with attributes:

ap_class_index: array([0, 1, 2, 3])
box: ultralytics.utils.metrics.Metric object
confusion_matrix: <ultralytics.utils.metrics.ConfusionMatrix object at 0x7849912a0590>
curves: ['Precision-Recall(B)', 'F1-Confidence(B)', 'Precision-Confidence(B)', 'Recall-Confidence(B)']
curves_results: [[array([          0,    0.001001,    0.002002,    0.003003,    0.004004,    0.005005,    0.006006,    0.007007,    0.008008,    0.009009,     0.01001,    0.011011,    0.012012,    0.013013,    0.014014,    0.015015,    0.016016,    0.017017,    0.018018,    0.019019,     0.02002,    0.021021,    0.022022,    0.023023,
          0.024024,    0.025025,    0.026026,    0.027027,    0.028028,    0.029029,     0.03003,    0.031031,    0.032032,    0.033033,    0.034034,    0.035035,    0.036036,    0.037037,    0.038038,    0.039039,     0.04004,    0.041041,    0.042042,    0.043043,    0.044044,    0.045045,    0.046046,    0.047047,
          0

### Konfigurasi Training Modifikasi (MobileNetV3 Tuning)

| Kategori | Parameter | Nilai | Deskripsi |
| :--- | :--- | :--- | :--- |
| **Utama** | `model` | yolov8n_mbv3.yaml | Arsitektur modifikasi MobileNetV3 |
| | `imgsz` | 800 | Resolusi input gambar (High Resolution) |
| | `epochs` | 50 | Total siklus pelatihan |
| | `batch` | 16 | Jumlah sampel per batch |
| **Optimizer** | `optimizer` | SGD | Stochastic Gradient Descent |
| | `lr0` | 0.01 | Initial Learning Rate |
| | `momentum` | 0.937 | Faktor momentum gradien |
| | `weight_decay` | 0.0005 | Penalti regularisasi |
| **Loss Gain** | `box` | 7.5 | Bobot penalti Box Loss |
| | `cls` | 0.5 | Bobot penalti Classification Loss |
| | `dfl` | 1.5 | Bobot Distribution Focal Loss |
| **Augmentasi** | `mosaic` | 1.0 | Aktif (Penggabungan 4 gambar) |
| | `close_mosaic`| 15 | Mosaic dimatikan pada 15 epoch terakhir |
| | `warmup_epochs`| 3 | Durasi fase pemanasan awal |
| **Lainnya** | `iou` | 0.7 | NMS Threshold (Penyaringan kotak dobel) |


In [ ]:
from ultralytics import YOLO

model_eval = YOLO('/content/runs/detect/yolo_tuning_test/yolov8n_mbv3_tuning/weights/best.pt')

# Lakukan validasi ulang secara singkat
print("\n--- Menghitung Metrik Lengkap ---")
metrik = model_eval.val(data='/content/drive/MyDrive/RDD2022_JP_YOLO/rdd_data_jp_local.yaml')

# Panggil metrik spesifik
print(f"mAP@50    : {metrik.box.map50:.4f}")
print(f"mAP@75    : {metrik.box.map75:.4f}")
print(f"mAP@50-95 : {metrik.box.map:.4f}")


--- Menghitung Metrik Lengkap ---
Ultralytics 8.4.48 🚀 Python-3.12.13 torch-2.10.0+cu128 CUDA:0 (Tesla T4, 14913MiB)
YOLOv8n_mbv3 summary: 215 layers, 3,006,367 parameters, 0 gradients, 6.1 GFLOPs
val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 2375.8±836.3 MB/s, size: 79.1 KB)
val: Scanning /content/rdd_data/content/drive/MyDrive/RDD2022_JP_YOLO/val/labels.cache... 1863 images, 277 backgrounds, 0 corrupt: 100% ━━━━━━━━━━━━ 1863/1863 558.1Mit/s 0.0s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 117/117 5.0it/s 23.5s
                   all       1863       3345      0.581      0.478      0.515      0.235
    longitudinal crack        571        844      0.564      0.418      0.465      0.204
      transverse crack        446        806      0.512      0.371      0.397      0.146
       alligator crack        940       1233      0.664       0.62       0.67      0.353
               pothole        297        462      0.584  

##**TUNING - 2**

In [ ]:
from ultralytics import YOLO

print("\n--- Memulai Training Final: Modifikasi + Partial Freezing ---")
model = YOLO('yolov8n_mbv3.yaml')

model.train(
    data='/content/drive/MyDrive/RDD2022_JP_YOLO/rdd_data_jp_local.yaml',
    epochs=50,
    imgsz=640,
    batch=32,
    freeze=10,
    lr0=0.01,
    optimizer='SGD',
    label_smoothing=0.1,
    name='yolov8n_mbv3_tuning_frozen',
    project='yolo_tuning_test'
)


--- Memulai Training Final: Modifikasi + Partial Freezing ---


WARNING ⚠️ 'label_smoothing' is deprecated and will be removed in the future.
Ultralytics 8.4.48 🚀 Python-3.12.13 torch-2.10.0+cu128 CUDA:0 (Tesla T4, 14913MiB)
engine/trainer: agnostic_nms=False, amp=True, angle=1.0, augment=False, auto_augment=randaugment, batch=32, bgr=0.0, box=7.5, cache=False, cfg=None, classes=None, close_mosaic=10, cls=0.5, cls_pw=0.0, compile=False, conf=None, copy_paste=0.0, copy_paste_mode=flip, cos_lr=False, cutmix=0.0, data=/content/drive/MyDrive/RDD2022_JP_YOLO/rdd_data_jp_local.yaml, degrees=0.0, deterministic=True, device=None, dfl=1.5, dnn=False, dropout=0.0, dynamic=False, embed=None, end2end=None, epochs=50, erasing=0.4, exist_ok=False, fliplr=0.5, flipud=0.0, format=torchscript, fraction=1.0, freeze=10, half=False, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, imgsz=640, int8=False, iou=0.7, keras=False, kobj=1.0, line_width=None, lr0=0.01, lrf=0.01, mask_ratio=4, max_det=300, mixup=0.0, mode=train, model=yolov8n_mbv3.yaml, momentum=0.937, mosaic=1.0, multi_sca

  0                  -1  1    927008  MobileNetV3Backbone                          []                            
  1                   0  1         0  Select_Explicit_Fix                          [3, 24, 0]                    
  2                   0  1         0  Select_Explicit_Fix                          [3, 48, 1]                    
  3                   0  1         0  Select_Explicit_Fix                          [3, 576, 2]                   
  4                  -1  1    461888  ultralytics.nn.modules.block.SPPF            [576, 256, 5]                 
  5                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']          
  6             [-1, 2]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
  7                  -1  1     44416  ultralytics.nn.modules.block.C2f             [304, 64, 1]                  
  8                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None

ultralytics.utils.metrics.DetMetrics object with attributes:

ap_class_index: array([0, 1, 2, 3])
box: ultralytics.utils.metrics.Metric object
confusion_matrix: <ultralytics.utils.metrics.ConfusionMatrix object at 0x79b305e0e5d0>
curves: ['Precision-Recall(B)', 'F1-Confidence(B)', 'Precision-Confidence(B)', 'Recall-Confidence(B)']
curves_results: [[array([          0,    0.001001,    0.002002,    0.003003,    0.004004,    0.005005,    0.006006,    0.007007,    0.008008,    0.009009,     0.01001,    0.011011,    0.012012,    0.013013,    0.014014,    0.015015,    0.016016,    0.017017,    0.018018,    0.019019,     0.02002,    0.021021,    0.022022,    0.023023,
          0.024024,    0.025025,    0.026026,    0.027027,    0.028028,    0.029029,     0.03003,    0.031031,    0.032032,    0.033033,    0.034034,    0.035035,    0.036036,    0.037037,    0.038038,    0.039039,     0.04004,    0.041041,    0.042042,    0.043043,    0.044044,    0.045045,    0.046046,    0.047047,
          0

### Konfigurasi Training Modifikasi (Partial Freezing)

| Kategori | Parameter | Nilai | Deskripsi |
| :--- | :--- | :--- | :--- |
| **Utama** | `model` | yolov8n_mbv3.yaml | Arsitektur modifikasi MobileNetV3 |
| | `imgsz` | 640 | Resolusi input gambar |
| | `epochs` | 50 | Total siklus pelatihan |
| | `batch` | 32 | Jumlah sampel per batch (Batch lebih besar) |
| **Optimasi** | `optimizer` | SGD | Stochastic Gradient Descent |
| | `lr0` | 0.01 | Initial Learning Rate |
| | `freeze` | 10 | Membekukan 10 layer pertama (Partial Freeze) |
| | `warmup_epochs`| 3.0 | Durasi fase pemanasan awal |
| **Loss Gain** | `box` | 7.5 | Bobot penalti Box Loss |
| | `cls` | 0.5 | Bobot penalti Classification Loss |
| | `dfl` | 1.5 | Bobot Distribution Focal Loss |
| **Augmentasi** | `mosaic` | 1.0 | Aktif (Penggabungan 4 gambar) |
| | `close_mosaic`| 10 | Mosaic dimatikan pada 10 epoch terakhir |
| **Lainnya** | `iou` | 0.7 | NMS Threshold (Penyaringan kotak dobel) |

In [ ]:
from ultralytics import YOLO

model_eval = YOLO('/content/runs/detect/yolo_tuning_test/yolov8n_mbv3_tuning_frozen-2/weights/best.pt')

# Lakukan validasi ulang secara singkat
print("\n--- Menghitung Metrik Lengkap ---")
metrik = model_eval.val(data='/content/drive/MyDrive/RDD2022_JP_YOLO/rdd_data_jp_local.yaml')

# Panggil metrik spesifik
print(f"mAP@50    : {metrik.box.map50:.4f}")
print(f"mAP@75    : {metrik.box.map75:.4f}")
print(f"mAP@50-95 : {metrik.box.map:.4f}")


--- Menghitung Metrik Lengkap ---
Ultralytics 8.4.48 🚀 Python-3.12.13 torch-2.10.0+cu128 CUDA:0 (Tesla T4, 14913MiB)
YOLOv8n_mbv3 summary: 215 layers, 3,006,367 parameters, 0 gradients, 6.1 GFLOPs
val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 1964.3±536.8 MB/s, size: 74.7 KB)
val: Scanning /content/rdd_data/content/drive/MyDrive/RDD2022_JP_YOLO/val/labels.cache... 1863 images, 277 backgrounds, 0 corrupt: 100% ━━━━━━━━━━━━ 1863/1863 601.1Mit/s 0.0s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 117/117 5.5it/s 21.3s
                   all       1863       3345      0.409      0.368      0.342      0.148
    longitudinal crack        571        844      0.413      0.341      0.306      0.124
      transverse crack        446        806      0.346      0.169      0.177     0.0592
       alligator crack        940       1233      0.464      0.557      0.515      0.254
               pothole        297        462      0.414  

##**TUNING - 3**

In [ ]:
from ultralytics import YOLO

model = YOLO('yolov8n_mbv3.yaml')

model.train(
    data='/content/drive/MyDrive/RDD2022_JP_YOLO/rdd_data_jp_local.yaml',
    epochs=50,
    imgsz=800,
    batch=16,
    optimizer='AdamW',
    lr0=0.001,
    close_mosaic=15,
    name='yolov8n_mbv3_tuning_2',
    project='yolo_tuning_test'
)

model.safetensors:   0%|          | 0.00/10.2M [00:00<?, ?B/s]

Ultralytics 8.4.48 🚀 Python-3.12.13 torch-2.10.0+cu128 CUDA:0 (Tesla T4, 14913MiB)
engine/trainer: agnostic_nms=False, amp=True, angle=1.0, augment=False, auto_augment=randaugment, batch=16, bgr=0.0, box=7.5, cache=False, cfg=None, classes=None, close_mosaic=15, cls=0.5, cls_pw=0.0, compile=False, conf=None, copy_paste=0.0, copy_paste_mode=flip, cos_lr=False, cutmix=0.0, data=/content/drive/MyDrive/RDD2022_JP_YOLO/rdd_data_jp_local.yaml, degrees=0.0, deterministic=True, device=None, dfl=1.5, dnn=False, dropout=0.0, dynamic=False, embed=None, end2end=None, epochs=50, erasing=0.4, exist_ok=False, fliplr=0.5, flipud=0.0, format=torchscript, fraction=1.0, freeze=None, half=False, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, imgsz=800, int8=False, iou=0.7, keras=False, kobj=1.0, line_width=None, lr0=0.001, lrf=0.01, mask_ratio=4, max_det=300, mixup=0.0, mode=train, model=yolov8n_mbv3.yaml, momentum=0.937, mosaic=1.0, multi_scale=0.0, name=yolov8n_mbv3_tuning_2, nbs=64, nms=False, opset=None, optimize

  0                  -1  1    927008  MobileNetV3Backbone                          []                            
  1                   0  1         0  Select_Explicit_Fix                          [3, 24, 0]                    
  2                   0  1         0  Select_Explicit_Fix                          [3, 48, 1]                    
  3                   0  1         0  Select_Explicit_Fix                          [3, 576, 2]                   
  4                  -1  1    461888  ultralytics.nn.modules.block.SPPF            [576, 256, 5]                 
  5                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']          
  6             [-1, 2]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
  7                  -1  1     44416  ultralytics.nn.modules.block.C2f             [304, 64, 1]                  
  8                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None

ultralytics.utils.metrics.DetMetrics object with attributes:

ap_class_index: array([0, 1, 2, 3])
box: ultralytics.utils.metrics.Metric object
confusion_matrix: <ultralytics.utils.metrics.ConfusionMatrix object at 0x7e7f881bf770>
curves: ['Precision-Recall(B)', 'F1-Confidence(B)', 'Precision-Confidence(B)', 'Recall-Confidence(B)']
curves_results: [[array([          0,    0.001001,    0.002002,    0.003003,    0.004004,    0.005005,    0.006006,    0.007007,    0.008008,    0.009009,     0.01001,    0.011011,    0.012012,    0.013013,    0.014014,    0.015015,    0.016016,    0.017017,    0.018018,    0.019019,     0.02002,    0.021021,    0.022022,    0.023023,
          0.024024,    0.025025,    0.026026,    0.027027,    0.028028,    0.029029,     0.03003,    0.031031,    0.032032,    0.033033,    0.034034,    0.035035,    0.036036,    0.037037,    0.038038,    0.039039,     0.04004,    0.041041,    0.042042,    0.043043,    0.044044,    0.045045,    0.046046,    0.047047,
          0

### Konfigurasi Training Modifikasi (AdamW Tuning)

| Kategori | Parameter | Nilai | Deskripsi |
| :--- | :--- | :--- | :--- |
| **Utama** | `model` | yolov8n_mbv3.yaml | Arsitektur modifikasi MobileNetV3 |
| | `imgsz` | 800 | Resolusi input gambar (High Resolution) |
| | `epochs` | 50 | Total siklus pelatihan |
| | `batch` | 16 | Jumlah sampel per batch |
| **Optimizer** | `optimizer` | AdamW | Adaptive Moment Estimation with Weight Decay |
| | `lr0` | 0.001 | Initial Learning Rate (Lebih kecil untuk stabilitas) |
| | `momentum` | 0.937 | Beta1 untuk AdamW |
| | `weight_decay` | 0.0005 | Penalti regularisasi |
| **Loss Gain** | `box` | 7.5 | Bobot penalti Box Loss |
| | `cls` | 0.5 | Bobot penalti Classification Loss |
| | `dfl` | 1.5 | Bobot Distribution Focal Loss |
| **Augmentasi** | `mosaic` | 1.0 | Aktif (Penggabungan 4 gambar) |
| | `close_mosaic`| 15 | Mosaic dimatikan pada 15 epoch terakhir |
| | `warmup_epochs`| 3.0 | Durasi fase pemanasan awal |
| **Lainnya** | `iou` | 0.7 | NMS Threshold (Penyaringan kotak dobel) |

In [ ]:
from ultralytics import YOLO

model_eval = YOLO('/content/runs/detect/yolo_tuning_test/yolov8n_mbv3_tuning_2/weights/best.pt')

# Lakukan validasi ulang secara singkat
print("\n--- Menghitung Metrik Lengkap ---")
metrik = model_eval.val(data='/content/drive/MyDrive/RDD2022_JP_YOLO/rdd_data_jp_local.yaml')

# Panggil metrik spesifik
print(f"mAP@50    : {metrik.box.map50:.4f}")
print(f"mAP@75    : {metrik.box.map75:.4f}")
print(f"mAP@50-95 : {metrik.box.map:.4f}")


--- Menghitung Metrik Lengkap ---
Ultralytics 8.4.48 🚀 Python-3.12.13 torch-2.10.0+cu128 CUDA:0 (Tesla T4, 14913MiB)
YOLOv8n_mbv3 summary: 215 layers, 3,006,367 parameters, 0 gradients, 6.1 GFLOPs
val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 2544.2±793.7 MB/s, size: 79.1 KB)
val: Scanning /content/rdd_data/content/drive/MyDrive/RDD2022_JP_YOLO/val/labels.cache... 1863 images, 277 backgrounds, 0 corrupt: 100% ━━━━━━━━━━━━ 1863/1863 781.4Mit/s 0.0s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 117/117 4.6it/s 25.4s
                   all       1863       3345      0.547      0.489        0.5      0.227
    longitudinal crack        571        844      0.525      0.422      0.432      0.192
      transverse crack        446        806      0.497      0.366      0.374      0.135
       alligator crack        940       1233      0.625       0.63      0.664      0.347
               pothole        297        462      0.541  

##**TUNING - 4**

In [ ]:
from ultralytics import YOLO

model = YOLO('yolov8n_mbv3.yaml')

model.train(
    data='/content/drive/MyDrive/RDD2022_JP_YOLO/rdd_data_jp_local.yaml',
    epochs=50,
    imgsz=800,
    batch=16,
    optimizer='AdamW',
    lr0=0.001,
    box=15.0,           # Push akurasi lokasi retakan
    cls=2.0,            # Push akurasi klasifikasi tipe kerusakan
    close_mosaic=20,    # KUNCI UTAMA: 20 epoch terakhir fokus gambar bersih
    warmup_epochs=3,    # Percepat masuk ke fase belajar utama
    name='yolov8n_mbv3_tuning_3',
    project='yolo_tuning_test'
)

/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:93: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


model.safetensors:   0%|          | 0.00/10.2M [00:00<?, ?B/s]

Ultralytics 8.4.48 🚀 Python-3.12.13 torch-2.10.0+cu128 CUDA:0 (Tesla T4, 14913MiB)
engine/trainer: agnostic_nms=False, amp=True, angle=1.0, augment=False, auto_augment=randaugment, batch=16, bgr=0.0, box=15.0, cache=False, cfg=None, classes=None, close_mosaic=20, cls=2.0, cls_pw=0.0, compile=False, conf=None, copy_paste=0.0, copy_paste_mode=flip, cos_lr=False, cutmix=0.0, data=/content/drive/MyDrive/RDD2022_JP_YOLO/rdd_data_jp_local.yaml, degrees=0.0, deterministic=True, device=None, dfl=1.5, dnn=False, dropout=0.0, dynamic=False, embed=None, end2end=None, epochs=50, erasing=0.4, exist_ok=False, fliplr=0.5, flipud=0.0, format=torchscript, fraction=1.0, freeze=None, half=False, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, imgsz=800, int8=False, iou=0.7, keras=False, kobj=1.0, line_width=None, lr0=0.001, lrf=0.01, mask_ratio=4, max_det=300, mixup=0.0, mode=train, model=yolov8n_mbv3.yaml, momentum=0.937, mosaic=1.0, multi_scale=0.0, name=yolov8n_mbv3_tuning_3, nbs=64, nms=False, opset=None, optimiz

  0                  -1  1    927008  MobileNetV3Backbone                          []                            
  1                   0  1         0  Select_Explicit_Fix                          [3, 24, 0]                    
  2                   0  1         0  Select_Explicit_Fix                          [3, 48, 1]                    
  3                   0  1         0  Select_Explicit_Fix                          [3, 576, 2]                   
  4                  -1  1    461888  ultralytics.nn.modules.block.SPPF            [576, 256, 5]                 
  5                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']          
  6             [-1, 2]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
  7                  -1  1     44416  ultralytics.nn.modules.block.C2f             [304, 64, 1]                  
  8                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None

ultralytics.utils.metrics.DetMetrics object with attributes:

ap_class_index: array([0, 1, 2, 3])
box: ultralytics.utils.metrics.Metric object
confusion_matrix: <ultralytics.utils.metrics.ConfusionMatrix object at 0x7a82221bae40>
curves: ['Precision-Recall(B)', 'F1-Confidence(B)', 'Precision-Confidence(B)', 'Recall-Confidence(B)']
curves_results: [[array([          0,    0.001001,    0.002002,    0.003003,    0.004004,    0.005005,    0.006006,    0.007007,    0.008008,    0.009009,     0.01001,    0.011011,    0.012012,    0.013013,    0.014014,    0.015015,    0.016016,    0.017017,    0.018018,    0.019019,     0.02002,    0.021021,    0.022022,    0.023023,
          0.024024,    0.025025,    0.026026,    0.027027,    0.028028,    0.029029,     0.03003,    0.031031,    0.032032,    0.033033,    0.034034,    0.035035,    0.036036,    0.037037,    0.038038,    0.039039,     0.04004,    0.041041,    0.042042,    0.043043,    0.044044,    0.045045,    0.046046,    0.047047,
          0

### Konfigurasi Training Modifikasi (High Gain & Early Close)

| Kategori | Parameter | Nilai | Deskripsi |
| :--- | :--- | :--- | :--- |
| **Utama** | `model` | yolov8n_mbv3.yaml | Arsitektur modifikasi MobileNetV3 |
| | `imgsz` | 800 | Resolusi tinggi untuk detail retakan |
| | `epochs` | 50 | Total siklus pelatihan |
| | `batch` | 16 | Jumlah sampel per batch |
| **Optimasi** | `optimizer` | AdamW | Optimizer adaptif (Adam with Weight Decay) |
| | `lr0` | 0.001 | Initial Learning Rate stabil |
| **Loss Gain** | `box` | 15.0 | Bobot Box Loss dinaikkan 2x lipat (Agresif) |
| | `cls` | 2.0 | Bobot Class Loss dinaikkan 4x lipat |
| | `dfl` | 1.5 | Bobot Distribution Focal Loss |
| **Augmentasi** | `mosaic` | 1.0 | Aktif di awal training |
| | `close_mosaic`| 20 | Mosaic dimatikan lebih awal (20 epoch terakhir) |
| | `warmup_epochs`| 3 | Durasi fase pemanasan awal |
| **Lainnya** | `iou` | 0.7 | NMS Threshold |
| | `nc` | 4 | Jumlah kelas kerusakan jalan utama |

In [ ]:
from ultralytics import YOLO

model_eval = YOLO('/content/runs/detect/yolo_tuning_test/yolov8n_mbv3_tuning_3/weights/best.pt')

# Lakukan validasi ulang secara singkat
print("\n--- Menghitung Metrik Lengkap ---")
metrik = model_eval.val(data='/content/drive/MyDrive/RDD2022_JP_YOLO/rdd_data_jp_local.yaml')

# Panggil metrik spesifik
print(f"mAP@50    : {metrik.box.map50:.4f}")
print(f"mAP@75    : {metrik.box.map75:.4f}")
print(f"mAP@50-95 : {metrik.box.map:.4f}")


--- Menghitung Metrik Lengkap ---
Ultralytics 8.4.48 🚀 Python-3.12.13 torch-2.10.0+cu128 CUDA:0 (Tesla T4, 14913MiB)
YOLOv8n_mbv3 summary: 215 layers, 3,006,367 parameters, 0 gradients, 6.1 GFLOPs
val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 1713.0±421.7 MB/s, size: 79.1 KB)
val: Scanning /content/rdd_data/content/drive/MyDrive/RDD2022_JP_YOLO/val/labels.cache... 1863 images, 277 backgrounds, 0 corrupt: 100% ━━━━━━━━━━━━ 1863/1863 488.4Mit/s 0.0s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 117/117 4.7it/s 24.9s
                   all       1863       3345      0.572      0.467      0.498      0.225
    longitudinal crack        571        844       0.57      0.383      0.429      0.187
      transverse crack        446        806      0.499      0.371      0.381      0.135
       alligator crack        940       1233      0.657      0.607      0.659      0.344
               pothole        297        462      0.561  

##**TUNING - 5**

In [ ]:
from ultralytics import YOLO

model = YOLO('yolov8n_mbv3.yaml')

model.train(
    data='/content/drive/MyDrive/RDD2022_JP_YOLO/rdd_data_jp_local.yaml',
    epochs=50,
    imgsz=800,
    batch=16,
    optimizer='AdamW',
    lr0=0.01,
    cos_lr=True,
    close_mosaic=20,
    name='yolov8n_mbv3_tuning_4',
    project='yolo_tuning_test'
    )

Ultralytics 8.4.50 🚀 Python-3.12.13 torch-2.10.0+cu128 CUDA:0 (Tesla T4, 14913MiB)
engine/trainer: agnostic_nms=False, amp=True, angle=1.0, augment=False, auto_augment=randaugment, batch=16, bgr=0.0, box=7.5, cache=False, cfg=None, classes=None, close_mosaic=20, cls=0.5, cls_pw=0.0, compile=False, conf=None, copy_paste=0.0, copy_paste_mode=flip, cos_lr=True, cutmix=0.0, data=/content/drive/MyDrive/RDD2022_JP_YOLO/rdd_data_jp_local.yaml, degrees=0.0, deterministic=True, device=None, dfl=1.5, dnn=False, dropout=0.0, dynamic=False, embed=None, end2end=None, epochs=50, erasing=0.4, exist_ok=False, fliplr=0.5, flipud=0.0, format=torchscript, fraction=1.0, freeze=None, half=False, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, imgsz=800, int8=False, iou=0.7, keras=False, kobj=1.0, line_width=None, lr0=0.01, lrf=0.01, mask_ratio=4, max_det=300, mixup=0.0, mode=train, model=yolov8n_mbv3.yaml, momentum=0.937, mosaic=1.0, multi_scale=0.0, name=yolov8n_mbv3_tuning_4-2, nbs=64, nms=False, opset=None, optimize

  0                  -1  1    927008  MobileNetV3Backbone                          []                            
  1                   0  1         0  Select_Explicit_Fix                          [3, 24, 0]                    
  2                   0  1         0  Select_Explicit_Fix                          [3, 48, 1]                    
  3                   0  1         0  Select_Explicit_Fix                          [3, 576, 2]                   
  4                  -1  1    461888  ultralytics.nn.modules.block.SPPF            [576, 256, 5]                 
  5                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']          
  6             [-1, 2]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
  7                  -1  1     44416  ultralytics.nn.modules.block.C2f             [304, 64, 1]                  
  8                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None

ultralytics.utils.metrics.DetMetrics object with attributes:

ap_class_index: array([0, 1, 2, 3])
box: ultralytics.utils.metrics.Metric object
confusion_matrix: <ultralytics.utils.metrics.ConfusionMatrix object at 0x7f0b52501550>
curves: ['Precision-Recall(B)', 'F1-Confidence(B)', 'Precision-Confidence(B)', 'Recall-Confidence(B)']
curves_results: [[array([          0,    0.001001,    0.002002,    0.003003,    0.004004,    0.005005,    0.006006,    0.007007,    0.008008,    0.009009,     0.01001,    0.011011,    0.012012,    0.013013,    0.014014,    0.015015,    0.016016,    0.017017,    0.018018,    0.019019,     0.02002,    0.021021,    0.022022,    0.023023,
          0.024024,    0.025025,    0.026026,    0.027027,    0.028028,    0.029029,     0.03003,    0.031031,    0.032032,    0.033033,    0.034034,    0.035035,    0.036036,    0.037037,    0.038038,    0.039039,     0.04004,    0.041041,    0.042042,    0.043043,    0.044044,    0.045045,    0.046046,    0.047047,
          0

In [ ]:
from ultralytics import YOLO

model_eval = YOLO('/content/runs/detect/yolo_tuning_test/yolov8n_mbv3_tuning_4/weights/best.pt')

# Lakukan validasi ulang secara singkat
print("\n--- Menghitung Metrik Lengkap ---")
metrik = model_eval.val(data='/content/drive/MyDrive/RDD2022_JP_YOLO/rdd_data_jp_local.yaml')

# Panggil metrik spesifik
print(f"mAP@50    : {metrik.box.map50:.4f}")
print(f"mAP@75    : {metrik.box.map75:.4f}")
print(f"mAP@50-95 : {metrik.box.map:.4f}")

FileNotFoundError: [Errno 2] No such file or directory: '/content/runs/detect/yolo_tuning_test/yolov8n_mbv3_tuning_4/weights/best.pt'

##**TUNING - 6**

In [ ]:
from ultralytics import YOLO

model = YOLO('yolov8n_mbv3.yaml')

model.train(
    data='/content/drive/MyDrive/RDD2022_JP_YOLO/rdd_data_jp_local.yaml',
    epochs=50,
    imgsz=800,
    batch=16,
    optimizer='SGD',
    lr0=0.01,
    freeze=10,
    name='yolov8n_mbv3_tuning_5',
    project='yolo_tuning_test'
    )

/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:93: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


model.safetensors:   0%|          | 0.00/10.2M [00:00<?, ?B/s]

Ultralytics 8.4.50 🚀 Python-3.12.13 torch-2.10.0+cu128 CUDA:0 (Tesla T4, 14913MiB)
engine/trainer: agnostic_nms=False, amp=True, angle=1.0, augment=False, auto_augment=randaugment, batch=16, bgr=0.0, box=7.5, cache=False, cfg=None, classes=None, close_mosaic=10, cls=0.5, cls_pw=0.0, compile=False, conf=None, copy_paste=0.0, copy_paste_mode=flip, cos_lr=False, cutmix=0.0, data=/content/drive/MyDrive/RDD2022_JP_YOLO/rdd_data_jp_local.yaml, degrees=0.0, deterministic=True, device=None, dfl=1.5, dnn=False, dropout=0.0, dynamic=False, embed=None, end2end=None, epochs=50, erasing=0.4, exist_ok=False, fliplr=0.5, flipud=0.0, format=torchscript, fraction=1.0, freeze=10, half=False, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, imgsz=800, int8=False, iou=0.7, keras=False, kobj=1.0, line_width=None, lr0=0.01, lrf=0.01, mask_ratio=4, max_det=300, mixup=0.0, mode=train, model=yolov8n_mbv3.yaml, momentum=0.937, mosaic=1.0, multi_scale=0.0, name=yolov8n_mbv3_tuning_5, nbs=64, nms=False, opset=None, optimize=Fa

  0                  -1  1    927008  MobileNetV3Backbone                          []                            
  1                   0  1         0  Select_Explicit_Fix                          [3, 24, 0]                    
  2                   0  1         0  Select_Explicit_Fix                          [3, 48, 1]                    
  3                   0  1         0  Select_Explicit_Fix                          [3, 576, 2]                   
  4                  -1  1    461888  ultralytics.nn.modules.block.SPPF            [576, 256, 5]                 
  5                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']          
  6             [-1, 2]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
  7                  -1  1     44416  ultralytics.nn.modules.block.C2f             [304, 64, 1]                  
  8                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None

ultralytics.utils.metrics.DetMetrics object with attributes:

ap_class_index: array([0, 1, 2, 3])
box: ultralytics.utils.metrics.Metric object
confusion_matrix: <ultralytics.utils.metrics.ConfusionMatrix object at 0x7cc78dc6b6b0>
curves: ['Precision-Recall(B)', 'F1-Confidence(B)', 'Precision-Confidence(B)', 'Recall-Confidence(B)']
curves_results: [[array([          0,    0.001001,    0.002002,    0.003003,    0.004004,    0.005005,    0.006006,    0.007007,    0.008008,    0.009009,     0.01001,    0.011011,    0.012012,    0.013013,    0.014014,    0.015015,    0.016016,    0.017017,    0.018018,    0.019019,     0.02002,    0.021021,    0.022022,    0.023023,
          0.024024,    0.025025,    0.026026,    0.027027,    0.028028,    0.029029,     0.03003,    0.031031,    0.032032,    0.033033,    0.034034,    0.035035,    0.036036,    0.037037,    0.038038,    0.039039,     0.04004,    0.041041,    0.042042,    0.043043,    0.044044,    0.045045,    0.046046,    0.047047,
          0

In [ ]:
from ultralytics import YOLO

model_eval = YOLO('/content/runs/detect/yolo_tuning_test/yolov8n_mbv3_tuning_5/weights/best.pt')

# Lakukan validasi ulang secara singkat
print("\n--- Menghitung Metrik Lengkap ---")
metrik = model_eval.val(data='/content/drive/MyDrive/RDD2022_JP_YOLO/rdd_data_jp_local.yaml')

# Panggil metrik spesifik
print(f"mAP@50    : {metrik.box.map50:.4f}")
print(f"mAP@75    : {metrik.box.map75:.4f}")
print(f"mAP@50-95 : {metrik.box.map:.4f}")


--- Menghitung Metrik Lengkap ---
Ultralytics 8.4.50 🚀 Python-3.12.13 torch-2.10.0+cu128 CUDA:0 (Tesla T4, 14913MiB)
YOLOv8n_mbv3 summary: 215 layers, 3,006,367 parameters, 0 gradients, 6.1 GFLOPs
val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 2482.5±842.6 MB/s, size: 79.1 KB)
val: Scanning /content/rdd_data/content/drive/MyDrive/RDD2022_JP_YOLO/val/labels.cache... 1863 images, 277 backgrounds, 0 corrupt: 100% ━━━━━━━━━━━━ 1863/1863 651.2Mit/s 0.0s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 117/117 4.4it/s 26.6s
                   all       1863       3345      0.389      0.389       0.35      0.151
    longitudinal crack        571        844      0.397      0.379       0.35      0.147
      transverse crack        446        806      0.319       0.19      0.175     0.0567
       alligator crack        940       1233      0.413      0.576      0.505      0.243
               pothole        297        462      0.428  

##**NYIMPEN KE GDRIVE**

In [ ]:
import shutil
import os

sumber_runs = '/content/runs'
tujuan_drive = '/content/drive/MyDrive/RDD2022_JP_YOLO/Hasil_Training_Skripsi'

if os.path.exists(sumber_runs):
    shutil.copytree(sumber_runs, tujuan_drive, dirs_exist_ok=True)
    print(f" Selesai! Folder 'runs' sudah aman di: {tujuan_drive}")
else:
    print(" Error: Folder 'runs' tidak ditemukan. Pastikan proses training sudah berjalan.")

 Selesai! Folder 'runs' sudah aman di: /content/drive/MyDrive/RDD2022_JP_YOLO/Hasil_Training_Skripsi


In [ ]:
import shutil
import os

nama_folder_training = 'yolov8n_mbv3_tuning_4'
folder_project = 'yolo_tuning_test'
# ------------------------------------------------

sumber_skenario = f'/content/runs/detect/{folder_project}/{nama_folder_training}'
tujuan_drive = f'/content/drive/MyDrive/RDD2022_JP_YOLO/Hasil_Training_Skripsi/Tuning_Manual/{nama_folder_training}'

if os.path.exists(sumber_skenario):
    os.makedirs(tujuan_drive, exist_ok=True)
    shutil.copytree(sumber_skenario, tujuan_drive, dirs_exist_ok=True)

    print(f"Backup Berhasil!")
    print(f"Folder '{nama_folder_training}' sudah aman di Drive.")
    print(f"Path: {tujuan_drive}")
else:
    print(f" Error: Folder {sumber_skenario} tidak ditemukan.")

Backup Berhasil!
Folder 'yolov8n_mbv3_tuning_5' sudah aman di Drive.
Path: /content/drive/MyDrive/RDD2022_JP_YOLO/Hasil_Training_Skripsi/Tuning_Manual/yolov8n_mbv3_tuning_5
